### PR TITLE
diff_working_copies: Simplify tree checkout functions

### PR DESCRIPTION
**Pull request by isuffix**
This commit will be affected by changes to executable bit behavior, but this is separate enough that it warrants its own PR.

# Checklist

If applicable:

- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [ ] I have added/updated tests to cover my changes

### CODE REVIEW

File: cli/src/merge_tools/external.rs [hunk 1]
```diff
@@ -440,11 +445,14 @@ pub fn generate_diff(
         conflict_marker_style,
     };
     let store = left_tree.store();
-    let diff_wc = check_out_trees(store, left_tree, right_tree, matcher, None, &options)?;
-    set_readonly_recursively(diff_wc.left_working_copy_path())
-        .map_err(ExternalToolError::SetUpDir)?;
-    set_readonly_recursively(diff_wc.right_working_copy_path())
-        .map_err(ExternalToolError::SetUpDir)?;
+    let diff_wc = check_out_trees(
```
Comment (yuja) [inline cli/src/merge_tools/external.rs:448]: Here we should make both left and right sides readonly. It's not super important, though.

I think `set_readonly_recursively()` can be handled by caller because it's not a job of `check_out_trees()`.

File: cli/src/merge_tools/diff_working_copies.rs [hunk 2]
```diff
@@ -137,20 +107,24 @@ pub(crate) fn set_readonly_recursively(path: &Path) -> Result<(), std::io::Error
...
+///
+/// TwoWay: prepare a left and right tree; left is readonly.
+/// ThreeWay: prepare left, right, and output trees; left + right are readonly.
 #[derive(Debug, Clone, Copy)]
-pub(crate) enum DiffSide {
-    // Left,
-    Right,
+pub(crate) enum DiffType {
+    TwoWay,
+    ThreeWay,
```
Comment (yuja) [inline cli/src/merge_tools/diff_working_copies.rs]: nit: I think the original name `DiffSide` is also good. `DiffType` has various meanings.

File: cli/src/merge_tools/external.rs (hunk 1 above)
Comment (isuffix) [reply cli/src/merge_tools/external.rs:448]: We could change the name of the function, but I believe moving them to `check_out_trees()` makes it easier for the caller to do the right thing. Note that I did update the function's comment to mention this behavior.

I think this is helpful because it centralizes two places that were doing the same operation before, which each had to peer inside of the returned value of `check_out_trees()` to do so, and which now simply pass the value on.

File: cli/src/merge_tools/diff_working_copies.rs (hunk 2 above)
Comment (isuffix) [reply cli/src/merge_tools/diff_working_copies.rs]: How about `DiffWays` or `DiffKind`? Though if you're set on it, I'm fine with `DiffSide`.

File: cli/src/merge_tools/external.rs (hunk 1 above)
Comment (yuja) [reply cli/src/merge_tools/external.rs:448]: If we can keep the function arguments sane, that's good. `check_out_trees()` already has lots of arguments, so I don't think it's good to add more for `set_readonly`. The choice of sides to be made readonly is arbitrary.

If you want to delete `<side>_working_copy_path()` for some reason, maybe you can add `DiffWorkingCopies::set_<side>_readonly()`?

File: cli/src/merge_tools/diff_working_copies.rs (hunk 2 above)
Comment (yuja) [reply cli/src/merge_tools/diff_working_copies.rs]: or `DiffSides` (plural)? `DiffKind` is basically the same as `DiffType`. `DiffWays` seems more informative, but it's a bit redundant that both type and variant names have `Side`.

File: cli/src/merge_tools/external.rs (hunk 1 above)
Comment (isuffix) [reply cli/src/merge_tools/external.rs:448]: The `DiffType` argument is already communicating that to the function, as explained by the doc comment on the enum:
```rs
/// How to prepare tree states from the working copy for a diff viewer/editor.
///
/// TwoWay: prepare a left and right tree; left is readonly.
/// ThreeWay: prepare left, right, and output trees; left + right are readonly.
```

File: cli/src/merge_tools/diff_working_copies.rs (hunk 2 above)
Comment (isuffix) [reply cli/src/merge_tools/diff_working_copies.rs]: To me, `DiffSide(s)` is just not correct with the new enum variant names. `TwoWay` and `ThreeWay` aren't referring to sides of the diff. They refer to the method of diffing, or more plainly: **the type of diff being performed.**

If this is just a nitpick as noted, then I'd prefer to leave it as `DiffType`. I don't think the name is actually important, and I don't think users are likely to confuse its meaning given the limited scope.

File: cli/src/merge_tools/external.rs (hunk 1 above)
Comment (yuja) [reply cli/src/merge_tools/external.rs:448]: And it breaks `generate_diff()`, right? It's a `TwoWay` diff of readonly left/right sides. It's also up to caller to make which side of `TwoWay` diff readonly. We don't have a caller who edits the left side right now, though.

Comment (isuffix) [reply cli/src/merge_tools/external.rs:448]: Apologies. I realize that I misread your original comment above, and did not understand that the current behavior is different than before. Given that, I now agree that `DiffWorkingCopy` should have methods for `set_<side>_readonly()`, and I will implement that.

**Comment by isuffix** [comment]
Is there anything else I need to do for this?

File: cli/src/merge_tools/diff_working_copies.rs [hunk 3]
```diff
@@ -160,48 +142,36 @@ pub(crate) fn check_out_trees(
...
-                },
-                changed_files,
-                options,
-            )
-        })
-        .transpose()?;
+    let temp_path = temp_dir.path();
+
+    // Checkout a tree into our temp directory with the given name prefix.
+    let check_out = |name: &str, tree, files| -> Result<TreeState, DiffCheckoutError> {
```
Comment (yuja) [inline cli/src/merge_tools/diff_working_copies.rs]: nit: `files` can be omitted since all snapshots should respect the same sparse patterns.

File: cli/src/merge_tools/diff_working_copies.rs [hunk 4]
```diff
@@ -160,48 +142,36 @@ pub(crate) fn check_out_trees(
...
+        let wc_path = temp_path.join(name);
+        let state_dir = temp_path.join(format!("{name}_state"));
+        std::fs::create_dir(&wc_path).map_err(DiffCheckoutError::SetUpDir)?;
+        std::fs::create_dir(&state_dir).map_err(DiffCheckoutError::SetUpDir)?;
+        let mut state = TreeState::init(
+            store.clone(),
+            wc_path.clone(),
```
Comment (yuja) [inline cli/src/merge_tools/diff_working_copies.rs]: nit: redundant `wc_path.clone()`

File: cli/src/merge_tools/diff_working_copies.rs [hunk 5]
```diff
@@ -160,48 +142,36 @@ pub(crate) fn check_out_trees(
...
+        state.check_out(tree, options)?;
+        Ok(state)
+    };
+
+    let left = check_out("left", left_tree, changed_files.clone())?;
+    let output = match diff_type {
+        DiffType::TwoWay => None,
+        DiffType::ThreeWay => Some(check_out("output", right_tree, changed_files.clone())?),
+    };
+    let right = check_out("right", right_tree, changed_files)?;
```
Comment (yuja) [inline cli/src/merge_tools/diff_working_copies.rs]: nit: check out in `left`, `right`, `output` order for consistency?

File: cli/src/merge_tools/diff_working_copies.rs [hunk 6]
```diff
@@ -218,40 +188,52 @@ impl DiffEditWorkingCopies {
...
-                File::create(&output_instructions_path).map_err(ExternalToolError::SetUpDir)?;
-            if diff_wc.right_working_copy_path() != output_wc_path {
-                let mut right_instructions_file =
-                    File::create(diff_wc.right_working_copy_path().join("JJ-INSTRUCTIONS"))
-                        .map_err(ExternalToolError::SetUpDir)?;
-                right_instructions_file
-                    .write_all(
-                        b"\
+        if output_instructions_path.exists() {
+            return Ok(output_instructions_path);
```
Comment (yuja) [inline cli/src/merge_tools/diff_working_copies.rs]: should return `Ok(None)`? If the instruction path conflicts with the tree content, the tree path shouldn't be deleted on cleanup.

Can you also add test for this edge case?

File: cli/src/merge_tools/diff_working_copies.rs (hunk 5 above)
Comment (joyously) [reply cli/src/merge_tools/diff_working_copies.rs]: or a comment explaining the order? Or explaining what it is consistent with?

File: cli/src/merge_tools/diff_working_copies.rs (hunk 4 above)
Comment (isuffix) [reply cli/src/merge_tools/diff_working_copies.rs]: Ah, yep. Thanks

File: cli/src/merge_tools/diff_working_copies.rs (hunk 3 above)
Comment (isuffix) [reply cli/src/merge_tools/diff_working_copies.rs]: The original reason is to avoid an extra clone of `changed_files`, which is the same reason as the comment below. I'll mark this as resolved to focus on that comment. If we do decide we're fine with the clone, then I'll be sure to remove the parameter in the closure.

File: cli/src/merge_tools/diff_working_copies.rs (hunk 5 above)
Comment (isuffix) [reply cli/src/merge_tools/diff_working_copies.rs]: The reason for this was simply to avoid an unnecessary clone of `changed_files`. The problem is that `output` can't go last because it may or may not exist, and so can't easily take ownership of the vector. If you're ok with the extra clone, I can just clone all three times and reorder them to be consistent. Otherwise I can add a comment and/or reorder them like `output`, `right`, `left` which is a little better.

Comment (isuffix) [reply cli/src/merge_tools/diff_working_copies.rs]: Another option is refactoring to look like this? I don't know if it helps that much, but it is convenient and clear
```rust
let output = match diff_type {
    DiffType::TwoWay => None,
    DiffType::ThreeWay => Some(check_out("output", right_tree, changed_files.clone())?),
};
Ok(DiffWorkingCopies {
    left: check_out("left", left_tree, changed_files.clone())?,
    right: check_out("left", left_tree, changed_files)?,
    output,
    _temp_dir: temp_dir,
})
```

File: cli/src/merge_tools/diff_working_copies.rs (hunk 6 above)
Comment (isuffix) [reply cli/src/merge_tools/diff_working_copies.rs]: Ah, good catch. Thank you.

File: cli/src/merge_tools/diff_working_copies.rs (hunk 5 above)
Comment (yuja) [reply cli/src/merge_tools/diff_working_copies.rs]: Extra `changed_files.clone()` wouldn't matter. The `check_out()` process would be way slower than `vec.clone()`.

Comment (isuffix) [reply cli/src/merge_tools/diff_working_copies.rs]: Ok, then I'll clone inside the function

File: cli/src/merge_tools/diff_working_copies.rs [hunk 7]
```diff
@@ -160,48 +142,36 @@ pub(crate) fn check_out_trees(
...
+
+    // Checkout a tree into our temp directory with the given prefix.
+    let check_out = |name: &str, tree| -> Result<TreeState, DiffCheckoutError> {
+        let wc_path = temp_path.join(name);
+        let state_dir = temp_path.join(format!("{name}_state"));
+        std::fs::create_dir(&wc_path).map_err(DiffCheckoutError::SetUpDir)?;
+        std::fs::create_dir(&state_dir).map_err(DiffCheckoutError::SetUpDir)?;
+        let mut state = TreeState::init(
+            store.clone(),
+            wc_path.clone(),
```
Comment (yuja) [inline cli/src/merge_tools/diff_working_copies.rs]: nit: `wc_path.clone()` can be omitted
